agent/logger/global.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package logger
+
+import (
+	"sync"
+
+	"github.com/cihub/seelog"
+)
+
+var (
+	loggerMux              = &sync.RWMutex{}
+	globalSeelog           seelog.LoggerInterface
+	globalStructuredLogger *structuredLogger
+)
+
+func init() {
+	loggerMux.Lock()
+	defer loggerMux.Unlock()
+	if globalSeelog == nil {
+		globalSeelog = seelog.Current
+		globalStructuredLogger = newStructuredLogger(globalSeelog, DEFAULT_OUTPUT_FORMAT)
+	}
+}
+
+func setGlobalLogger(li seelog.LoggerInterface, logFormat string) {
+	loggerMux.Lock()
+	defer loggerMux.Unlock()
+	globalSeelog = li
+	globalStructuredLogger = newStructuredLogger(li, logFormat)
+}
+
+func getGlobalStructuredLogger() *structuredLogger {
+	loggerMux.RLock()
+	defer loggerMux.RUnlock()
+	return globalStructuredLogger
+}
+
+func Trace(message string, fields ...Fields) {
+	l := getGlobalStructuredLogger()
+	l.Trace(message, fields...)
+}
+
+func Debug(message string, fields ...Fields) {
+	l := getGlobalStructuredLogger()
+	l.Debug(message, fields...)
+}
+
+func Info(message string, fields ...Fields) {
+	l := getGlobalStructuredLogger()
+	l.Info(message, fields...)
+}
+
+func Warn(message string, fields ...Fields) error {
+	l := getGlobalStructuredLogger()
+	return l.Warn(message, fields...)
+}
+
+func Error(message string, fields ...Fields) error {
+	l := getGlobalStructuredLogger()
+	return l.Error(message, fields...)
+}
+
+func Critical(message string, fields ...Fields) error {
+	l := getGlobalStructuredLogger()
+	return l.Critical(message, fields...)
+}
+
+func Close() {
```

Comment:
nit: should we also add `Close` (and the other functions below) to `StructuredLogger` interface? That way we can abstract away `seelog` from the global logger, as well as make the functions more consistent

Reply:
Yeah, I'l add that to the final revision of this 👍🏼

Reply:
ended up not doing this yet, because turns out that Close is invoked internally by seelog, as you can see in the UTs